src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,139 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private T[] arrayList;
+    private int size = 0;
+    private int emptyPlaces = 10;
+    private final double ARRAY_ENLARGEMENT_PERCENTAGE = 0.5;
+
+    public ArrayList() {
+        arrayList = (T[]) new Object[emptyPlaces];
+    }
+
+    public void enlarge() {
+        emptyPlaces = (int) (size * ARRAY_ENLARGEMENT_PERCENTAGE);
+        T[] temporaryEnlargedArrayList = (T[]) new Object[size + emptyPlaces];
+        for (int i = 0; i < size; i++) {
+            temporaryEnlargedArrayList[i] = arrayList[i];
+        }
+        arrayList = (T[]) new Object[size + emptyPlaces];
+        for (int i = 0; i < size; i++) {
+            arrayList[i] = temporaryEnlargedArrayList[i];
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (emptyPlaces == 0) {
+            enlarge();
+        }
+        arrayList[size] = value;
+        emptyPlaces--;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("given index does not exist (add by index)");
+        } else {
+            if (index == size) {
+                add(value);
+            } else {
+                T[] temporaryArrayList = (T[]) new Object[size - index];
+                for (int i = index; i < size; i++) {
+                    temporaryArrayList[i - index] = arrayList[i];
+                }
+
+                set(value, index);
+
+                for (int i = 0; i < temporaryArrayList.length - 1; i++) {
```

Comment:
This loop and the use of 'set' method is unnecessary. You should use System.arraycopy() to handle the insertion and shifting of elements.